src/main/java/duke/Storage.java

```diff
@@ -0,0 +1,112 @@
+package duke;
+
+import duke.exception.DukeDatabaseException;
+import duke.task.Deadline;
+import duke.task.Event;
+import duke.task.Task;
+import duke.task.TaskList;
+import duke.task.Todo;
+
+import java.io.BufferedWriter;
+import java.io.File;
+import java.io.FileWriter;
+import java.io.FileNotFoundException;
+import java.io.IOException;
+import java.time.LocalDateTime;
+import java.time.format.DateTimeFormatter;
+import java.util.ArrayList;
+import java.util.Scanner;
+
+
+/**
+ * Represents the storage that loads data from and saves data to the database file.
+ */
+public class Storage {
+    private String filePath;
+
+    /**
+     * Constructor for Storage.
+     *
+     * @param filePath the path to the txt database file
+     */
+    public Storage(String filePath) {
+        this.filePath = filePath;
+    }
+
+
+    /**
+     * Load data from the txt database file.
+     *
+     * @return an array of tasks as represented by the database
+     * @throws DukeDatabaseException if database could not be created due to unknown error
+     */
+    public Task[] loadData() throws DukeDatabaseException {
+        File db = new File(this.filePath);
+        ArrayList<Task> taskList = new ArrayList<>();
+        try {
+            Scanner fileReader = new Scanner(db);
+            while (fileReader.hasNextLine()) {
+                taskList.add(readEntry(fileReader.nextLine()));
+            }
+            fileReader.close();
+        } catch (FileNotFoundException e) {
+            this.createDatabase();
+        }
+        return taskList.toArray(new Task[0]);
+    }
+
+
+    private void createDatabase() throws DukeDatabaseException {
+        File db = new File(this.filePath);
+        File dir = new File(db.getParent());
+        dir.mkdir();
+        try {
+            db.createNewFile();
+        } catch (IOException e) {
+            throw new DukeDatabaseException();
+        }
+    }
+
+
+    private Task readEntry(String entry) {
+        String[] fields = entry.split("\\|");
+        Task taskToAdd;
+        DateTimeFormatter formatter = DateTimeFormatter.ofPattern("yyyy-MM-dd HH:mm");
+
+        switch (fields[0]) {
+        case "T":
+            taskToAdd = new Todo(fields[2]);
+            break;
+        case "E":
+            taskToAdd = new Event(fields[2],
+                    LocalDateTime.parse(fields[3], formatter),
+                    LocalDateTime.parse(fields[4], formatter));
+            break;
+        default: // (case "D")
```

Comment:
Hmm, can we assume that the todo type in the data file is always valid? Perhaps we could explicitly still have a case for D and use default for potential erroneous data!

Reply:
Yep I should've throw a `DukeDatabaseException` for the default case here instead!